File: src/core/graph_utils.py
import heapq
import networkx as nx
from typing import Optional, Tuple, Dict, List
from .data_structures import Path, GraphState

def reverse(graph: nx.DiGraph) -> nx.DiGraph:
    Gr = nx.DiGraph()
    Gr.add_edges_from((v,u,d) for u,v,d in graph.edges(data=True))
    return Gr


def dijkstra(
    graph: nx.DiGraph, 
    src: int, 
    dest: int
    ) -> Optional[Path]:    
    
    if src == dest:
        path = Path()
        path.route = [src]
        return path

    heap = [(0, src, [])]
    visited = set()

    while heap:
        cost, node, path_list = heapq.heappop(heap)

        if node in visited:
            continue
        visited.add(node)

        if node == dest:
            shortest_path = Path()

            for i in range(len(path_list)):
                if i < len(path_list) - 1:
                    u, v = path_list[i], path_list[i+1]
                    shortest_path.edges[(u, v)] = graph[u][v]['weight']
                    shortest_path.length += graph[u][v]['weight']
                    shortest_path.route.append(u)
                else:
                    u, v = path_list[i], dest
                    shortest_path.edges[(u, v)] = graph[u][v]['weight']
                    shortest_path.length += graph[u][v]['weight']
                    shortest_path.route.append(u)
                    shortest_path.route.append(v)

            shortest_path.lb = shortest_path.length
            return shortest_path

        for neighbor, data in graph[node].items():
            if neighbor not in visited:
                new_cost = cost + data['weight']
                heapq.heappush(heap, (new_cost, neighbor, path_list + [node]))

    return None

def construct_partial_spt(graph_state: GraphState, v: int) -> float:
    if graph_state.isSettled[v]:
        return graph_state.distances[v]

    while graph_state.PQ:
        cost, node = heapq.heappop(graph_state.PQ)

        if cost > graph_state.distances[node]:
            continue

        if not graph_state.isSettled[node]:
            graph_state.isSettled[node] = True

            for neighbor, data in graph_state.graph_reverse[node].items():
                if not graph_state.isSettled[neighbor]:
                    new_cost = cost + data['weight']

                    if new_cost < graph_state.distances[neighbor]:
                        graph_state.distances[neighbor] = new_cost
                        graph_state.parent[neighbor] = node
                        heapq.heappush(graph_state.PQ, (new_cost, neighbor))

            if node == v:
                return graph_state.distances[v]

    return float('inf')


def _build_path(
    graph: nx.DiGraph, 
    path_list: List[int], 
    dest: int, 
    cost: float
) -> Path:
    shortest_path = Path()
    
    for i in range(len(path_list)):
        if i < len(path_list) - 1:
            u, v = path_list[i], path_list[i + 1]
        else:
            u, v = path_list[i], dest
        
        shortest_path.edges[(u, v)] = graph[u][v]['weight']
        shortest_path.length += graph[u][v]['weight']
        shortest_path.route.append(u)
    
    shortest_path.route.append(dest)
    shortest_path.lb = shortest_path.length
    return shortest_path

File: src/main/java/org/example/core/GraphUtils.java
package org.example.core;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

import org.example.core.Graph.NodeDistance;

public class GraphUtils {
    public static Graph reverse(Graph graph) {
        Graph graphReverse = new Graph();

        for (int u : graph.getNodes()) {
            for (int v : graph.getNeighbors(u)) {
                Double weight = graph.getEdgeWeight(u, v);
                graphReverse.addWeightedEdge(v, u, weight);
            }
        }

        return graphReverse;
    }

    public static Path dijkstra(Graph graph, int src, int dest) {
        if (!graph.containsNode(src) || !graph.containsNode(dest)) {
            throw new IllegalArgumentException("Source ya da destination grafte yok");
        }

        if (src == dest) {
            Path path = new Path();
            path.setRoute(new ArrayList<>(Arrays.asList(src)));
            path.setLength(0.0);
            return path;
        }

        Map<Integer, Double> distances = new HashMap<>();
        Map<Integer, Integer> previous = new HashMap<>();
        PriorityQueue<NodeDistance> pq = new PriorityQueue<>(
            Comparator.comparingDouble(NodeDistance::getDistance)
        );

        for (int node : graph.getNodes()) {
            distances.put(node, Double.POSITIVE_INFINITY);
            previous.put(node, null);
        }
        distances.put(src, 0.0);
        pq.offer(new NodeDistance(src, 0.0));

        while (!pq.isEmpty()) {
            NodeDistance current = pq.poll();
            int u = current.getNode();

            if (current.getDistance() > distances.get(u)) {
                continue;
            }

            if (u == dest) {
                break;
            }

            for (int v : graph.getNeighbors(u)) {
                Double weight = graph.getEdgeWeight(u, v);
                double alt = distances.get(u) + weight;

                if (alt < distances.get(v)) {
                    distances.put(v, alt);
                    previous.put(v, u);
                    pq.offer(new NodeDistance(v, alt));
                }
            }
        }

        Path path = new Path();
        if (distances.get(dest) == Double.POSITIVE_INFINITY) {
            return path; 
        }

        List<Integer> route = new ArrayList<>();
        Integer current = dest;
        while (current != null) {
            route.add(0, current);
            current = previous.get(current);
        }

        path.setRoute(route);
        for (int i = 0; i < route.size() - 1; i++) {
            int u = route.get(i);
            int v = route.get(i + 1);
            Double weight = graph.getEdgeWeight(u, v);
            if (weight != null) {
                path.addEdge(u, v, weight);
            }
        }
        path.setLb(path.getLength());

        return path;
    }

    public static double constructPartialSPT(GraphState graphState, int v) {
        if (graphState.isSettled(v)) {
            return graphState.getDistance(v);
        }

        Graph graphReverse = graphState.getGraphReverse();

        while (!graphState.getPq().isEmpty()) {
            NodeDistance current = graphState.pollPQ();
            int node = current.getNode();
            double cost = current.getDistance();

            if (cost > graphState.getDistance(node)) {
                continue;
            }

            if (!graphState.isSettled(node)) {
                graphState.setSettled(node, true);

                for (int neighbor : graphReverse.getNeighbors(node)) {
                    if (!graphState.isSettled(neighbor)) {
                        Double weight = graphReverse.getEdgeWeight(node, neighbor);
                        if (weight != null) {
                            double newCost = cost + weight;

                            if (newCost < graphState.getDistance(neighbor)) {
                                graphState.setDistance(neighbor, newCost);
                                graphState.setParent(neighbor, node);
                                graphState.addToPQ(neighbor, newCost);
                            }
                        }
                    }
                }

                if (node == v) {
                    return graphState.getDistance(v);
                }
            }
        }

        return Double.POSITIVE_INFINITY;
    }

}
